docs/routing.md

```diff
@@ -0,0 +1,254 @@
+
+A simplistic overview of the routing process is as follows.
+
+- Call the handler with the request.
+- Check each phase from the build output configuration.
+  - Check each source route in a phase to see if it matches.
+  - If a match is found, apply any relevant response modifiers.
+  - Determine what to do next; return the match, or continue through routing.
+- Run the matched function, or fetch the matched static asset.
+- Return the final response.
+
+In order to prevent runaway phase checking, i.e. infinitely looping through phases, we have checks that determine whether we have processed phases too many times. This is indicative of records in the build output configuration causing infinite loops, and the counter resets each time the router is run for a request.
+
+### Checking Phases
+
+When we receive a request, the router looks for matches in the different phases, starting at the `none` phase.
+
+For every single source route in a phase, we check to see whether the route is a match, and process the relevant details (for more details on source checking see the [Checking Source Routes section](#checking-source-routes)). When we receive a response from the route checker, if we encountered an error — normally from running middleware — we break out of the routing process and return the error response. Otherwise, the only other time we disable the continuation of routing is if the route checker returns a final match.
+
+After all the source routes in the current phase have been checked, we move on to determine what will happen next. If the current phase is the `hit` phase, or the route has at some point been rewritten to an URL, the routing is complete and we're ready to server the the response.
+
+If the current phase is `miss`, we check the build output, if there is no file matching the requested one we set the status code to 404. Afterwards, if a file was found in the output, or the current phase is `error`, we move on to the `hit` phase. It is important that the `hit` phase is run every time a match is found in the build output or when the current phase is `miss` or `error`, because this allows us to update the headers for the response with the current matched path.
+
+In all other cases, we attempt to retrieve the next phase, based on the current one, and iterate forward.
+
+#### Diagram
+
+```mermaid
+flowchart TD
+    checkPhase[check phase] --for each route--> checkRoute
+    checkRoute[check route] -->|error| checkPhase
+    checkRoute -->|done| phaseHitMiss{is phase\nhit or miss?}
+    phaseHitMiss --> |phase == hit| matchingDone[matching done]
+    phaseHitMiss --> |phase == miss| set404[set 404]
+    phaseHitMiss --> |phase not neither| nextPhaseIsWhat{path found\nor\nphase == miss/error}
+    set404 --> nextPhaseIsWhat{path found\nsuccessfully?}
+    nextPhaseIsWhat --> |true\nset next phase to hit| checkPhase
+    nextPhaseIsWhat --> |false\nset next phase| checkPhase
+```
```

Comment:
```suggestion
```mermaid
flowchart TD
    checkPhase[check phase] --for each route--> checkRoute
    checkRoute[check route] -->|error| checkPhase
    checkRoute -->|done| phaseHitMiss{is phase\nhit or miss?}
    phaseHitMiss --> |phase == hit| matchingDone[matching done]
    phaseHitMiss --> |phase == miss| set404[set 404]
    phaseHitMiss --> |phase not neither| nextPhaseIsWhat{path found\nor\nphase == miss/error}
    set404 --> nextPhaseIsWhat{path found\nsuccessfully?}
    nextPhaseIsWhat --> |yes\nset next phase to hit| checkPhase
    nextPhaseIsWhat --> |no\nset next phase| checkPhase
```
```
sorry I just noticed this after seeing the other diagram 🙇